### ISSUE TITLE
WixSharp with v4 toolset not getting compiled and msi not found

### ISSUE THREAD

**Issue by abskulkarni** [oleg-shilo/wixsharp#1484]
Hi
I am trying to make Wix v4 work with WixSharp. 
I added a "WixSharp custom UI (Wix v4)" project as shown below and just compiled it without any change.

![image](https://github.com/oleg-shilo/wixsharp/assets/23351437/ced7d120-f44e-46af-9969-e663283a78b4)

![image](https://github.com/oleg-shilo/wixsharp/assets/23351437/da77bb87-d350-42f7-863d-c0b454def30a)

**Query 1:** It is not getting compiled. Compile error says "`wix.exe` cannot be found. Ensure you installed it with `dotnet tool install --global wix`" What does this mean? It is due to the post build event command?

![image](https://github.com/oleg-shilo/wixsharp/assets/23351437/1f776744-40de-4d6d-8675-42f5e68197be)

**Query 2:** If I remove post build even command, project gets compiled but there is no trace where msi file is generated. What is the issue here?

**Comment by oleg-shilo** [comment]
It means exactly what the error message says. 
You need to have `wix.exe` (WiX toolset) installed. 
It is a set of compilers, linkers and other WiX tools that WixSharp is using to orchestrate the building of msi files.

You can install wix.exe from the command prompt with this command: `dotnet tool install --global wix`

After that, you will be able to build your msi.

The post-build event there is for a reason. If you remove it then you will only compile the build script assembly but it will not be executed (you removed the execution command). So your msi will not be built.
You can read a bit more about this here: https://github.com/oleg-shilo/wixsharp/wiki/VS2019-%E2%80%93-2022-Templates#msi-authoring-steps

**Comment by abskulkarni** [comment]
Thanks @oleg-shilo , I did what is suggested by you above. It generated msi.
But now when I ran that msi, it gave me run time error "object reference not set..."

![image](https://github.com/oleg-shilo/wixsharp/assets/23351437/25d90990-bdd7-4a23-ac04-31e1029ec67c)



See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.NullReferenceException: Object reference not set to an instance of an object.
   at FIS.Installer.IDRCommandLine.Dialogs.WelcomeDialog.ResetLayout()
   at FIS.Installer.IDRCommandLine.Dialogs.WelcomeDialog.WelcomeDialog_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4645.0 built by: NET48REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework/v4.0.30319/mscorlib.dll
----------------------------------------
Microsoft.Deployment.WindowsInstaller
    Assembly Version: 3.0.0.0
    Win32 Version: 3.14.0.6526
    CodeBase: file:///C:/Users/E5370240/AppData/Local/Temp/2/MSI5811/Microsoft.Deployment.WindowsInstaller.DLL
----------------------------------------
WixSharp.UI
    Assembly Version: 1.25.2.0
    Win32 Version: 1.25.2.0
    CodeBase: file:///C:/Users/E5370240/AppData/Local/Temp/2/MSI5811/WixSharp.UI.DLL
----------------------------------------
WixSharp
    Assembly Version: 1.25.2.0
    Win32 Version: 1.25.2.0
    CodeBase: file:///C:/Users/E5370240/AppData/Local/Temp/2/MSI5811/WixSharp.DLL
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4654.0 built by: NET48REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4697.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4390.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4697.0 built by: NET48REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
FIS.Installer.IDRCommandLine
    Assembly Version: 1.0.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///C:/Users/E5370240/AppData/Local/Temp/2/MSI5811/FIS.Installer.IDRCommandLine.EXE
----------------------------------------
System.Xml.Linq
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml.Linq/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.Linq.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4084.0 built by: NET48REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.4190.0 built by: NET48REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by oleg-shilo** [comment]
Your `ResetLayout` method throws the exception because of something there being null while the reset was attempted.

Did you modify anything? Code, dialogs?

You have a few troubleshooting options here:
- You can debug your dialog code.
  Put an assertion statement and attach the debugger when it asserts ([see here](https://github.com/oleg-shilo/wixsharp/wiki/Tips'n'Tricks#debugging))
- Comment out `ResetLayout` completely. It's optional (just to improve scaling and rendering) and if there is something that prevents your dialogs from executing this method, then you can simply skip it.   
- Start with a VS simple template without custom UI. Just to verify that the MSI is built correctly.

**Comment by abskulkarni** [comment]
> Your `ResetLayout` method throws the exception because of something there being null while the reset was attempted.
> 
> Did you modify anything? Code, dialogs?
> 
> You have a few troubleshooting options here:
> 
> * You can debug your dialog code.
>   Put an assertion statement and attach the debugger when it asserts ([see here](https://github.com/oleg-shilo/wixsharp/wiki/Tips'n'Tricks#debugging))
> * Comment out `ResetLayout` completely. It's optional (just to improve scaling and rendering) and if there is something that prevents your dialogs from executing this method, then you can simply skip it.
> * Start with a VS simple template without custom UI. Just to verify that the MSI is built correctly.

I didn't change any code or dialog. I tried to build & run as it is. Also debugger is not getting attached to the process somehow. I know the process but it is not happening as expected.

**Comment by abskulkarni** [comment]
Hi @oleg-shilo 
I had to do two things as of now to build any WiX4 project successfully and see msi running.

1. Install wix.exe using command "dotnet tool install --global wix"
2. Create any WiX4 project. It gave me compile error for missing WixToolset.UI.wixext.dll in users .wix extension folder. So I installed "wixtoolset.ui.wixext v4.0.5" NuGet package in WiX4 project. Copied NuGet package's WixToolset.UI.wixext.dll to C:\Users\[User]\.wix\extensions\WixToolset.UI.wixext\5.0.0-rc.2\wixext4 folder which was empty after step 1.

I hope I have done right and will not get any unwanted surprises while developing WiX 4 projects here after

**Comment by oleg-shilo** [comment]
Great. You solved it.

A few things to comment... 

> Also debugger is not getting attached...

Usually, it is caused by the debugger not being elevated. Try to have your VS run as admin.
![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/fa6d46ec-2fd0-4761-943c-77f62f246c05)

> Install wix.exe using command "dotnet tool install --global wix"

Yep, that is a one-off step to install the required tools.

> So I installed "wixtoolset.ui.wixext v4.0.5"...

OMG, I cannot believe it. WiX team released a broken nuget package (5.0.0-rc) and it is still not patched. 
Please read this [thread](https://github.com/oleg-shilo/wixsharp/issues/1473) it discusses the workaround.

Basically, you need to delete v5.0.0-rc manually and then install the working version. Until WiX team fixes the problem (it's been more than 3 weeks now) you will have to use @VietBuiHa's [workaround](https://github.com/oleg-shilo/wixsharp/issues/1473#issuecomment-1987764824):

```
dotnet tool install --global wix --version 4.0.4
wix extension add -g WixToolset.UI.wixext/4.0.4
```
